README.md

```diff
@@ -1 +1,679 @@
+
+First, to explain why we will develop a little application that reads the content of a file. Every line in the file represents a specific entry from different sources, with the keys defining the system. We are only interested in the entries from `ELEMENT`. Those entries have the format `ELEMENT[00]=[00]`. The numbers directly after ELEMENT represent the order of the element. The numbers behind the equal sign are a number > 0, illustrating the ceiling-rounded elapsed percentage since the last entry. So, if `ELEMENT01=10` and `ELEMENT02=6`, the total percentage is 16, and the average is 8. As the percentage is ceiling rounded, the total number might go up to over 100%. As the overall system is highly parallel, but for some reason, everybody writes into this one file, it is locked most of the time, and the order of entries is not guaranteed.
+
+Our first task is to create a friendly process display on the console. It should show the percentage, the message "Started..." until it has reached 25%, the message "Running..." until it's at 75%, and "Almost done..." until it's at 100. Once it has reached 100 or more, we should show "Done" and no percentage. Also, we should make sure it's sorted - based on the last few entries, as a future requirement will be to calculate the time remaining.
+
+If you are already familiar with pipes, binding and higher-order functions, please skip the f# chapter and jump directly to [pipes and javascript](#pipes-and-javascript). If you are aware of the proposal and are only here to check what can go wrong, jump ahead to [the final chapter](#then-why-is-it-wrong).
+
+## Pipes and f#
+
+The pipe operator in f# is `|>`. It is defined as
+
+```fsharp
+value |> function = function value
+```
+
+Thus, it passes the value from the beginning to its end. Accordingly, if the function has another argument, it will be kept and the new element added at the end:
+
+```fsharp
+arg2 |> function arg1 = function arg1 arg2 
+```
+
+This is an easy thing to do for f#, as all functions are curried by default. That means that the function:
+
+```fsharp
+function arg1 arg2 
+```
+
+is really
+
+```fsharp
+function1 = fun arg1 -> fun arg2 
+```
+
+This is nothing you have to think about; the language does it automatically. In addition, you can also use an extension of the pipe to pass multiple arguments by providing a tuple and adding more `|`:
+
+```fsharp
+(arg1, arg2) ||> function = function arg1 arg2
+```
+
+Let's see how we can use that. We start with some test data and read it line by line:
+
+```fsharp
+let test_data = "
+ELEMENT01=3
+ELEMENT04=4
+discarded:incorrect
+ELEMENT03=1
+foo=4
+element06=12
+ELEMENT05=12
+bar=3
+ELEMENT02=12
+"
+
+let readLines(v: string) = v.Split[|'\n'|]
```

Comment:
Is there a reason for using snake_case for `test_data` but camelCase for `readLines`?